### PR TITLE
control/controlclient: start fetching the server noise key

### PR DESCRIPTION
**Pull request by maisem**
From https://github.com/tailscale/tailscale/pull/4062#issuecomment-1060961980

Updates https://github.com/tailscale/tailscale/issues/3488
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: control/controlclient/direct.go [hunk 1]
```diff
@@ -950,29 +956,38 @@ func encode(v interface{}, serverKey key.MachinePublic, mkey key.MachinePrivate)
...
+		return nil, fmt.Errorf("create control key request: %v", err)
 	}
-	req = req.WithContext(ctx)
 	res, err := httpc.Do(req)
 	if err != nil {
-		return key.MachinePublic{}, fmt.Errorf("fetch control key: %v", err)
+		return nil, fmt.Errorf("fetch control key: %v", err)
 	}
 	defer res.Body.Close()
 	b, err := ioutil.ReadAll(io.LimitReader(res.Body, 1<<16))
```
Comment (bradfitz) [inline control/controlclient/direct.go:970]: 64<<10

File: control/controlclient/direct.go [hunk 2]
```diff
@@ -950,29 +956,38 @@ func encode(v interface{}, serverKey key.MachinePublic, mkey key.MachinePrivate)
...
+
+	// Some old control servers might not be updated to send the new format.
+	// Fallback to the old way just in case.
 	k, err := key.ParseMachinePublicUntyped(mem.B(b))
 	if err != nil {
-		return key.MachinePublic{}, fmt.Errorf("fetch control key: %v", err)
+		return nil, multierr.New(jsonErr, err)
 	}
-	return k, nil
+	out.LegacyPublicKey = k
```
Comment (bradfitz) [inline control/controlclient/direct.go:989]: `out = tailcfg.OverTLSPublicKeyResponse{}` at line 982

File: control/controlclient/direct.go (hunk 1 above)
Comment (maisem) [reply control/controlclient/direct.go:970]: Oops, I squashed in the wrong commit. Will fix

**Comment by bradfitz** [comment]
Add the issue to the commit message.

File: control/controlclient/direct.go [hunk 3]
```diff
@@ -299,6 +302,8 @@ func (c *Direct) doLogin(ctx context.Context, opt loginOpt) (mustRegen bool, new
 	hi := c.hostinfo.Clone()
 	backendLogID := hi.BackendLogID
 	expired := c.expiry != nil && !c.expiry.IsZero() && c.expiry.Before(c.timeNow())
+	serverURL := c.serverURL
+	httpc := c.httpc
```
Comment (bradfitz) [inline control/controlclient/direct.go:306]: `c.httpc` isn't guarded by `c.mu` so this is misleading. Just go back to passing `c.httpc` on line 334 like before.

File: tstest/integration/testcontrol/testcontrol.go [hunk 4]
```diff
@@ -211,30 +215,38 @@ func (s *Server) serveUnhandled(w http.ResponseWriter, r *http.Request) {
...
+	s.noisePubKey = s.noisePrivKey.Public()
+	s.privKey = key.NewControl()
+	s.pubKey = s.privKey.Public()
 }
 
 func (s *Server) serveKey(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "text/plain")
 	w.WriteHeader(200)
-	io.WriteString(w, s.publicKey().UntypedHexString())
+	noiseKey, legacyKey := s.publicKey()
```
Comment (bradfitz) [inline tstest/integration/testcontrol/testcontrol.go:245]: no need to support old clients? we might want that later for integration tests with old clients.

I'd probably do the old thing here when r.FormValue("v") == ""

Comment (maisem) [reply tstest/integration/testcontrol/testcontrol.go]: Done

File: control/controlclient/direct.go (hunk 3 above)
Comment (maisem) [reply control/controlclient/direct.go]: Done

File: control/controlclient/direct.go (hunk 2 above)
Comment (maisem) [reply control/controlclient/direct.go:989]: Done

File: control/controlclient/direct.go [hunk 5]
```diff
@@ -299,6 +302,7 @@ func (c *Direct) doLogin(ctx context.Context, opt loginOpt) (mustRegen bool, new
 	hi := c.hostinfo.Clone()
 	backendLogID := hi.BackendLogID
 	expired := c.expiry != nil && !c.expiry.IsZero() && c.expiry.Before(c.timeNow())
+	serverURL := c.serverURL
```
Comment (bradfitz) [inline control/controlclient/direct.go:305]: also looks like this isn't guarded by c.mu.

File: tstest/integration/testcontrol/testcontrol.go [hunk 6]
```diff
@@ -211,30 +215,42 @@ func (s *Server) serveUnhandled(w http.ResponseWriter, r *http.Request) {
...
 	}
-	return s.pubKey, s.privKey
+	s.noisePrivKey = key.NewControl()
+	s.noisePubKey = s.noisePrivKey.Public()
+	s.privKey = key.NewControl()
+	s.pubKey = s.privKey.Public()
 }
 
 func (s *Server) serveKey(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "text/plain")
```
Comment (bradfitz) [inline tstest/integration/testcontrol/testcontrol.go:243]: wrong content-type for json mode. It's only text/plain for the old way, and application/json for the new way.

So move these two lines down into line 247 and 250.

You can drop the WriteHeader(200) actually in both cases. It's implicit.

Comment (maisem) [reply tstest/integration/testcontrol/testcontrol.go]: Done

File: control/controlclient/direct.go [hunk 7]
```diff
@@ -68,8 +69,10 @@ type Direct struct {
 	skipIPForwardingCheck  bool
 	pinger                 Pinger
 
-	mu           sync.Mutex // mutex guards the following fields
-	serverKey    key.MachinePublic
+	mu             sync.Mutex // mutex guards the following fields
+	serverKey      key.MachinePublic
```
Comment (bradfitz) [inline control/controlclient/direct.go:73]: `serverKey key.MachinePublic // original ("legacy") nacl crypto_box-based public key`

**Review by bradfitz — APPROVED** [review]

**Pull request merged**